inc/Search/Collection/FulltextCollection.php

```diff
@@ -0,0 +1,309 @@
+     * entity from the index
+     *
+     * @param string $entity the name of the entity
+     * @param string[] $tokens the list of tokens for this entity
+     *
+     * @throws IndexAccessException
+     * @throws IndexWriteException
+     * @throws IndexLockException
+     * @author Andreas Gohr <andi@splitbrain.org>
+     * @author Tom N Harris <tnharris@whoopdedo.org>
+     */
+    public function addEntity($entity, $tokens)
+    {
+        if(!$this->isWritable) {
+            throw new IndexLockException('Indexes not locked. Forgot to call lock()?');
+        }
+
+        $entityIndex = $this->getEntityIndex();
+        $entityId = $entityIndex->accessCachedValue($entity);
+
+        $old = $this->getReverseAssignments($entity); // assumes a frequency of 0
+        $new = $this->getTokenFrequency($tokens); // the real frequencies
+
+        $frequencies = array_replace_recursive(
+            $old,
+            $new
+        );
+
+        // store word frequency
+        foreach (array_keys($frequencies) as $tokenLength) {
+            $freqIndex = $this->getFrequencyIndex($tokenLength);
+            foreach ($frequencies[$tokenLength] as $tokenId => $freq) {
+                $record = $freqIndex->retrieveRow($tokenId);
+                $record = TupleOps::updateTuple($record, $entityId, $freq); // frequency of 0 deletes
+                $freqIndex->changeRow($tokenId, $record);
+
+                if (isset($oldwords[$tokenLength][$tokenId])) {
+                    unset($oldwords[$tokenLength][$tokenId]);
+                }
+            }
+            $freqIndex->save();
+        }
+
+        // update reverse Index
+        $this->saveReverseAssignments($entity, $frequencies);
+    }
+
+    /**
+     * Maximum suffix for the token indexes (eg. max word length currently stored)
+     * @return int
+     */
+    public function getTokenIndexMaximum()
+    {
+        return $this->getTokenIndex('')->max(); // no suffix needed to access the maximum
+    }
+
+
+    /**
+     *
+     * TokenIDs assigned to the given Entity sorted by token length as stored in the reverse Index
+     *
+     * Returns an Array in the form [tokenLength => [TokenId => 0, ...], ...]. The fixed 0 ensures array structure
+     * compatibility with getTokenFrequency() and is used to remove no longer used tokens.
+     *
+     * @param string $entity
+     * @return array
+     * @throws IndexAccessException
+     * @throws IndexWriteException
+     */
+    public function getReverseAssignments($entity)
+    {
+        $pageIndex = $this->getEntityIndex();
+        $entityId = $pageIndex->accessCachedValue($entity);
+
+        $pageRevIndex = $this->getReverseIndex();
+        $record = $pageRevIndex->retrieveRow($entityId);
+
+        $result = [];
+        if ($record === '') {
+            return $result;
+        }
+
+        foreach (explode(':', $record) as $row) {
+            list($tokenLength, $tokenId) = explode('*', $row);
```

Comment:
```suggestion
            [$tokenLength, $tokenId] = explode('*', $row);
```

i think php 7.2 allows this